kiosk/kiosk_browser/proxy.py

```diff
@@ -0,0 +1,113 @@
+"""Monitor proxy changes and automatically apply changes in Qt application"""
+
+import dbus
+import logging
+import threading
+import urllib
+from PyQt5.QtNetwork import QNetworkProxy
+from dbus.mainloop.glib import DBusGMainLoop
+from gi.repository import GLib
+
+def parse_url(url):
+    if url.startswith('http://'):
+        parsed = urllib.parse.urlparse(url)
+    else:
+        parsed = urllib.parse.urlparse(f'http://{url}')
+
+    if parsed.hostname != None and parsed.port != None:
+        return parsed
+    else:
+        logging.warn(f"Hostname or port missing in proxy url")
+        return None
+
+def extract_manual_proxy(config):
+    """Extract manual proxy from dbus configuration
+
+    Example configuration:
+
+    dbus.Dictionary({dbus.String('Servers'): dbus.Array([dbus.String('http://localhost:1234')], signature=dbus.Signature('s'), variant_level=1), dbus.String('Excludes'): dbus.Array([], signature=dbus.Signature('s'), variant_level=1), dbus.String('Method'): dbus.String('manual', variant_level=1)}, signature=dbus.Signature('sv'), variant_level=1)"""
+
+    if 'Method' in config:
+        method = config['Method']
+        if method == 'direct':
+            return None
+        elif method == 'manual':
+            if 'Servers' in config:
+                servers = config['Servers']
+                if len(servers) >= 1:
+                    return parse_url(servers[0])
+    else:
+        return None
+
+def get_current_proxy(bus):
+    """Get current proxy from dbus
+
+    Return  the proxy of a ready or connected service."""
+
+    client = dbus.Interface(
+        bus.get_object('net.connman', '/'),
+        'net.connman.Manager')
+
+    connected_services = [s for s in client.GetServices() if s[1]['State'] in ['ready', 'online']]
```

Comment:
Will this make sure that `online` services are preferred? I think they should be.

Can we find the service picked as default gateway?

Reply:
I was wondering whether it was possible to have many services in `online` or `ready` state, it seems so according to `doc/manager-api.txt`:

> If at least one service is in "ready" state and no service is in "online" state, then it will indicate the "ready" state.

https://github.com/dividat/playos/pull/54/commits/eaf3ac2227e903ea4d71ec72de39d66f297677de

Reply:
It is indeed possible, and the situation is made worse by connman seemingly not using the proxy information it has to perform the online check.

I just tested and observed that a client whose MAC is blocked from direct Internet access in the router will remain in `ready`. I don't understand whether or not this is the intended behaviour (https://git.kernel.org/pub/scm/network/connman/connman.git/tree/src/wispr.c).

We may want to think about also preferring services that have a gateway configured (_online_ > _gateway_ > _ready_).

Reply:
Looking at [this mailing list entry](https://lists.01.org/hyperkitty/list/connman@lists.01.org/message/GLKCXNUXPPUOARE5JURQXDG43RUVETXV/) it seems that connman itself relies on pacrunner to detect the proxy for a connection when making the online check: 

> If ConnMan gets to 'online' state, that is only a hint to the user that every connection should now
be working. Even the proxy ones - provided that the applications then is able to use for example pacrunner as there might still be a proxy preventing regular non-http PAC aware applications from functioning properly. Normally nobody runs pacrunner, meaning that if the host needs a proxy, the connection will still stay at 'ready'.

https://git.kernel.org/pub/scm/network/connman/connman.git/tree/src/proxy.c#n91
https://git.kernel.org/pub/scm/network/connman/connman.git/tree/plugins/pacrunner.c#n360

Reply:
> Looking at this mailing list entry it seems that connman itself relies on pacrunner to detect the proxy for a connection when making the online check

If I understand it correctly: if we don’t run pacrunner, connections behind a proxy will remain in _ready_ state. But running pacrunner, it would still not be able to use authenticated proxies, so connections behind authenticated proxies would remain in _ready_ state.

> Can we find the service picked as default gateway?

Sorry, I missed this part.

> We may want to think about also preferring services that have a gateway configured (online > gateway > ready).

I think that makes sense.

Reply:
> If I understand it correctly: if we don’t run pacrunner, connections behind a proxy will remain in _ready_ state. But running pacrunner, it would still not be able to use authenticated proxies, so connections behind authenticated proxies would remain in _ready_ state.

True, that would seem to be a consequence of the way connman does things.

Reply:
From `doc/overview-api.txt`:

> Within Connection Manager many services can be connected at the same time and also have an IP assignment. However only one can have the default route. The service with the default route will always be sorted at the top of the list.

https://git.kernel.org/pub/scm/network/connman/connman.git/tree/doc/overview-api.txt#n116

It would have been great to have an explicit flag, but it seems the right solution is to pick the first _online_ or _ready_ service.

Reply:
Picking the default service in https://github.com/dividat/playos/pull/54/commits/65109fc312bc6f134bd3b2c9046e778b4f2f4c41

Reply:
Ah, so you had it right all along. 😬 

Cool!

Reply:
I don’t remember if I did it for this reason though! It missed some documentation that is now added. :)